test/tests/functional/pbs_job_array.py

```diff
@@ -287,3 +290,276 @@ def test_qdel_expired_subjob(self):
+            extend='x')
+        self.server.expect(
+            JOB, {'comment': (MATCH_RE, '.*Job run at.*and failed')}, subjid_2,
+            extend='x')
+
+    def test_multiple_server_restarts(self):
+        """
+        Test subjobs wont rerun after multiple server restarts
+        """
+        a = {'resources_available.ncpus': 1}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+        a = {'job_history_enable': 1}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+        j = Job(TEST_USER, attrs={
+            ATTR_J: '1-2', 'Resource_List.select': 'ncpus=1'})
+        j.set_sleep_time(100)
+        j_id = self.server.submit(j)
+        subjid_1 = j.create_subjob_id(j_id, 1)
+        a = {'job_state': 'R', 'run_count': 1}
+        self.server.expect(JOB, a, subjid_1, attrop=PTL_AND)
+        for _ in range(5):
+            self.kill_and_restart_svr()
+            self.server.expect(JOB, a, subjid_1, attrop=PTL_AND)
+
+    def test_job_array_history_duration(self):
+        """
+        Test that job array and subjobs are purged after history duration
+        """
+        a = {'resources_available.ncpus': 1}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+        a = {'job_history_enable': 1}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+        a = {'job_history_duration': 30}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+        j = Job(TEST_USER, attrs={
+            ATTR_J: '1-2', 'Resource_List.select': 'ncpus=1'})
+        j.set_sleep_time(5)
+        j_id = self.server.submit(j)
+        subjid_1 = j.create_subjob_id(j_id, 1)
+        subjid_2 = j.create_subjob_id(j_id, 2)
+        a = {'job_state': 'R', 'run_count': 1}
+        self.server.expect(JOB, a, subjid_1, attrop=PTL_AND)
+        self.server.delete(subjid_1, extend='force')
+        b = {'job_state': 'X'}
+        self.server.expect(JOB, b, subjid_1)
+        self.server.expect(JOB, a, subjid_2, attrop=PTL_AND)
+        self.server.expect(JOB, 'job_state', op=UNSET, id=subjid_1, interval=5)
+        self.server.expect(JOB, 'job_state', op=UNSET, id=subjid_2, interval=5)
+
+    def test_queue_deletion_after_terminated_subjob(self):
+        """
+        Test that queue can be deleted after the job array is
+        terminated and server is restarted.
+        """
+        a = {'resources_available.ncpus': 1}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+        j = Job(TEST_USER, attrs={
+            ATTR_J: '1-2', 'Resource_List.select': 'ncpus=1'})
+        j.set_sleep_time(5)
+        j_id = self.server.submit(j)
+        subjid_1 = j.create_subjob_id(j_id, 1)
+        a = {'job_state': 'R', 'run_count': 1}
+        self.server.expect(JOB, a, subjid_1, attrop=PTL_AND)
+        self.server.delete(subjid_1, extend='force')
+        self.kill_and_restart_svr()
+        self.server.delete(j_id, wait=True)
+        self.server.manager(MGR_CMD_DELETE, QUEUE, id='workq')
```

Comment:
I'm not exactly sure of the reason for this test.  You're basically testing that you can delete an empty queue.

Reply:
There was a bug because of which we were not able to delete the queue, In the case of after job deletion jobs were not purged from db . Even when qstat did not show them.